src/rhs_function.c

```diff
@@ -0,0 +1,498 @@
+      vl_vec_int[ii] = -ur_vec_int[ii] * dum2 - vr_vec_int[ii] * dum1;
+
+    } else if (bl == 0 && br == 1) {
+      // Update right values as it is a reflective boundary wall
+      hr_vec_int[ii] = hl_vec_int[ii];
+
+      PetscReal dum1 = Square(sn_vec_int[ii]) - Square(cn_vec_int[ii]);
+      PetscReal dum2 = 2.0 * sn_vec_int[ii] * cn_vec_int[ii];
+
+      ur_vec_int[ii] = ul_vec_int[ii] * dum1 - vl_vec_int[ii] * dum2;
+      vr_vec_int[ii] = -ul_vec_int[ii] * dum2 - vl_vec_int[ii] * dum1;
+    }
+  }
+
+  // Call Riemann solver (only Roe currently supported)
+  PetscCheck(rdy->config.riemann == RIEMANN_ROE, rdy->comm, PETSC_ERR_USER, "Invalid Riemann solver selected! (Only roe is supported)");
+  PetscCall(ComputeRoeFlux(num, hl_vec_int, hr_vec_int, ul_vec_int, ur_vec_int, vl_vec_int, vr_vec_int, sn_vec_int, cn_vec_int, flux_vec_int,
+                           amax_vec_int));
+
+  // Save the flux values in the Vec based by TS
+  for (PetscInt ii = 0; ii < mesh->num_internal_edges; ii++) {
+    PetscInt  iedge      = edges->internal_edge_ids[ii];
+    PetscInt  cellOffset = edges->cell_offsets[iedge];
+    PetscInt  l          = edges->cell_ids[cellOffset];
+    PetscInt  r          = edges->cell_ids[cellOffset + 1];
+    PetscReal edgeLen    = edges->lengths[iedge];
+
+    PetscReal hl = x_ptr[l * ndof + 0];
+    PetscReal hr = x_ptr[r * ndof + 0];
+    PetscReal bl = b_ptr[l];
+    PetscReal br = b_ptr[r];
+
+    *amax_value = fmax(*amax_value, amax_vec_int[ii]);
+
+    if (bl == 0 && br == 0) {
+      // Both, left and right cells are not reflective boundary walls
+      if (!(hr < tiny_h && hl < tiny_h)) {
+        PetscReal areal = cells->areas[l];
+        PetscReal arear = cells->areas[r];
+
+        for (PetscInt idof = 0; idof < ndof; idof++) {
+          if (cells->is_local[l]) f_ptr[l * ndof + idof] -= flux_vec_int[ii][idof] * edgeLen / areal;
+          if (cells->is_local[r]) f_ptr[r * ndof + idof] += flux_vec_int[ii][idof] * edgeLen / arear;
+        }
+      }
+
+    } else if (bl == 1 && br == 0) {
+      // Left cell is a reflective boundary wall and right cell is an internal cell
+
+      PetscReal arear = cells->areas[r];
+      for (PetscInt idof = 0; idof < ndof; idof++) {
+        if (cells->is_local[r]) f_ptr[r * ndof + idof] += flux_vec_int[ii][idof] * edgeLen / arear;
+      }
+
+    } else if (bl == 0 && br == 1) {
+      // Left cell is an internal cell and right cell is a reflective boundary wall
+
+      PetscReal areal = cells->areas[l];
+      for (PetscInt idof = 0; idof < ndof; idof++) {
+        if (cells->is_local[l]) f_ptr[l * ndof + idof] -= flux_vec_int[ii][idof] * edgeLen / areal;
+      }
+    }
+  }
+
+  // Restore vectors
+  PetscCall(VecRestoreArray(rdy->X_local, &x_ptr));
+  PetscCall(VecRestoreArray(F, &f_ptr));
+  PetscCall(VecRestoreArray(rdy->B_local, &b_ptr));
+
+  PetscFunctionReturn(0);
+}
+
+// computes RHS on boundary edges
```

Comment:
Similarly, update the description https://github.com/RDycore/toy-problems/blob/main/swe_roe/ex2b.c#L1683